File: app/src/main/java/com/spidchenko/week2task/repositories/SearchRequestRepository.java
package com.spidchenko.week2task.repositories;

import androidx.lifecycle.LiveData;

import com.spidchenko.week2task.db.dao.SearchRequestDao;
import com.spidchenko.week2task.db.models.SearchRequest;

import java.util.List;
import java.util.concurrent.Executor;

public class SearchRequestRepository {

    private final SharedPrefRepository mSharedPrefRepository;
    private final Executor mExecutor;
    private final SearchRequestDao mSearchRequestDao;
    private static volatile SearchRequestRepository sInstance;

    private SearchRequestRepository(final SearchRequestDao searchRequestDao,
                                    final SharedPrefRepository sharedPrefRepository,
                                    final Executor executor) {
        mSharedPrefRepository = sharedPrefRepository;
        mSearchRequestDao = searchRequestDao;
        mExecutor = executor;
    }

    public static SearchRequestRepository getInstance(final SearchRequestDao searchRequestDao,
                                                      final SharedPrefRepository sharedPrefRepository,
                                                      final Executor executor) {
        if (sInstance == null) {
            synchronized (SearchRequestRepository.class) {
                if (sInstance == null) {
                    sInstance = new SearchRequestRepository(searchRequestDao, sharedPrefRepository, executor);
                }
            }
        }
        return sInstance;
    }

    public void saveCurrentSearchInDb(final SearchRequest request) {
        mExecutor.execute(() -> mSearchRequestDao.addSearchRequest(request));
    }

    public LiveData<List<SearchRequest>> getSearchRequests() {
        int userId = mSharedPrefRepository.getUserId();
        return mSearchRequestDao.getAllSearchRequests(userId);
    }

}

File: app/src/main/java/com/spidchenko/week2task/repositories/SearchRequestRepository.kt
package com.spidchenko.week2task.repositories

import androidx.lifecycle.LiveData
import com.spidchenko.week2task.db.dao.SearchRequestDao
import com.spidchenko.week2task.db.models.SearchRequest
import java.util.concurrent.Executor

class SearchRequestRepository private constructor(
    private val mSearchRequestDao: SearchRequestDao,
    private val mSharedPrefRepository: SharedPrefRepository,
    private val mExecutor: Executor
) {
    fun saveCurrentSearchInDb(request: SearchRequest?) {
        mExecutor.execute { mSearchRequestDao.addSearchRequest(request) }
    }

    val searchRequests: LiveData<List<SearchRequest?>?>?
        get() {
            val userId = mSharedPrefRepository.userId
            return mSearchRequestDao.getAllSearchRequests(userId)
        }

    companion object {
        @Volatile
        private var sInstance: SearchRequestRepository? = null

        @JvmStatic
        fun getInstance(
            searchRequestDao: SearchRequestDao,
            sharedPrefRepository: SharedPrefRepository,
            executor: Executor
        ): SearchRequestRepository? {
            if (sInstance == null) {
                synchronized(SearchRequestRepository::class.java) {
                    if (sInstance == null) {
                        sInstance = SearchRequestRepository(
                            searchRequestDao,
                            sharedPrefRepository,
                            executor
                        )
                    }
                }
            }
            return sInstance
        }
    }
}